pkg/runner/runner.go

```diff
@@ -0,0 +1,205 @@
+		case <-tctx.Done():
+			return nil
+		case <-t.C:
+
+		}
+	}
+}
+
+// RunTests runs a series of compliance tests
+// to make sure that server performs as expected
+func (r *Runner) RunTests(ctx context.Context) error {
+	r.Info("Starting compliance tests.")
+	var errors []error
+	for _, tc := range []func(ctx context.Context) error{
+		r.SendOK,
+		r.SendEmptyMetric,
+		r.SendCorruptedMetric,
+		func(ctx context.Context) error { return r.SendBogusAuth(ctx, "") },
+		func(ctx context.Context) error { return r.SendBogusAuth(ctx, uuid.New()) },
+	} {
+		err := tc(ctx)
+		if err != nil {
+			errors = append(errors, err)
+		}
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// SendOK sends OK metric and expects the metric to succeed
+func (r *Runner) SendOK(ctx context.Context) error {
+	r.Debugf("[TEST] Sending OK request...")
+	err := r.Client.Send(ctx, metric.Metric{
+		AccountID: ValidTestAccountID,
+		UserID:    ValidTestUserID,
+		Timestamp: time.Now().UTC(),
+	})
+	if err != nil {
+		r.Errorf("[FAIL] Sending OK request.")
+		return trace.BadParameter("expected success, have received %q error from server", err)
+	}
+	r.Infof("[PASS] Sending OK request.")
+	return nil
+}
+
+// SendBogusAuth sends a request with bogus authentication and expects to fail
+func (r *Runner) SendBogusAuth(ctx context.Context, bearerToken string) error {
+	r.Debugf("[TEST] Sending invalid authentication with bearer auth token %q.", bearerToken)
+	client, err := client.New(client.Config{
+		URL:    r.Client.URL,
+		CACert: r.Client.CACert,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = client.Send(ctx, metric.Metric{
+		AccountID: ValidTestAccountID,
+		UserID:    ValidTestUserID,
+		Timestamp: time.Now().UTC(),
+	})
+	if err == nil {
+		r.Errorf("[FAIL] Sending bogus request.")
+		return trace.BadParameter("expected access denied HTTP error, have received OK from server")
+	}
+	if !trace.IsAccessDenied(err) {
+		r.Errorf("[FAIL] Sending bogus request.")
+		return trace.BadParameter("expected access denied HTTP error, have received %q error from server", err)
+	}
+	r.Infof("[PASS] Sending bogus request.")
+	return nil
+}
+
+// SendEmptyMetric sends empty metric and expects an error
+func (r *Runner) SendEmptyMetric(ctx context.Context) error {
+	r.Debugf("[TEST] Sending empty request...")
+	err := r.Client.Send(ctx, metric.Metric{})
+	if err == nil {
+		r.Errorf("[FAIL] Sending empty request.")
+		return trace.BadParameter("runner sent an empty metric and received OK response, we expected HTTP Bad request error.")
+	}
+	if !trace.IsBadParameter(err) {
+		log.Warningf("runner sent an empty metric and received an error %v, however bad request response would have been better", err)
+	}
+	r.Infof("[PASS] Sending empty request.")
+	return nil
+}
+
+// SendCorruptedMetric sends corrupted metric
+func (r *Runner) SendCorruptedMetric(ctx context.Context) error {
+	r.Debugf("[TEST] Sending corrupted request...")
+	err := r.Client.Send(ctx, metric.Metric{})
```

Comment:
This sends the "empty" request just like the function above (`SendEmptyMetric`), not "corrupted". Should it send something else instead?

Reply:
fixed